### PR TITLE
Additional support for Tenable.OT Assets, Events and Plugins

### PR DESCRIPTION
**Pull request by adamdelman-tenable**
# Description
* Refactor Tenable.OT session client
* Add support for more Tenable.OT asset details
* Add support for Tenable.OT Events
* Add support for Tenable.OT Plugins

## Type of change
- [X ] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Tested fetching and parsing of Tenable.OT schema objects.

* Python Version(s) Tested: 3.10, 3.7

### CODE REVIEW

File: tenable/ot/api.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
+"""
+Events
+======
+
+Methods described in this section relate to the events API.
+These methods can be accessed at ``TenableOT.events``.
+
```
Comment (YevgenyKyrychenko) [inline tenable/ot/api.py:7]: Also need to mention assets, plugins + details

File: tenable/ot/api.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+    def list(
+        self,
+        query_filter: Optional[dict] = None,
+        search: Optional[str] = None,
+        sort: Optional[List[dict]] = None,
+        start_at: Optional[str] = None,
+        limit: Optional[int] = 200,
+        **kwargs
+    ):
```
Comment (YevgenyKyrychenko) [inline tenable/ot/api.py:196]: Return value type?

**Review by YevgenyKyrychenko — CHANGES_REQUESTED** [review]

File: tenable/ot/api.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (adamdelman-tenable) [reply tenable/ot/api.py]: fixed

File: tenable/ot/schema/events.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    src_asset_group: list[GroupMember]
+    dst_asset_group: list[GroupMember]
+    schedule: list[GroupMember]
+    protocol_group: list[GroupMember]
+    port_group: list[GroupMember]
+    tag_group: list[GroupMember]
+    value_group: list[GroupMember]
+    rule_group: list[GroupMember]
+    # params: ExtraParametersUnion
+    #
```
[the remark this answers is not in the document]
Comment (adamdelman-tenable) [reply tenable/ot/schema/events.py]: fixed

File: tenable/ot/schema/events.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    title: str
+    level: str
+    disabled: bool
+    archived: bool
+    schema: str
+    continuous: bool
+    snapshot: bool
+    system: bool
+    key: str
+    eventTypeDetails: list[EventTypeDetails]
```
[the remark this answers is not in the document]
Comment (adamdelman-tenable) [reply tenable/ot/schema/events.py]: fixed

File: tenable/ot/graphql/schema/plugins.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    id = fields.Integer(required=True)
+    name = fields.String(required=True)
+    owner = fields.String(allow_none=True)
+    severity = fields.String(required=True)
+    source = fields.String(required=True)
+    totalAffectedAssets = fields.Integer(required=True, data_key="totalAffectedAssets")
+    vpr_score = fields.Float(allow_none=True, data_key="vprScore")
+    affectedAssets = fields.Nested('AssetListSchema')
+    details = fields.Nested(PluginDetailsSchema, required=False)
+
```
[the remark this answers is not in the document]
Comment (adamdelman-tenable) [reply tenable/ot/graphql/schema/plugins.py:64]: fixed

File: tenable/ot/graphql/iterators.py [hunk 6]
```diff
@@ -24,28 +30,34 @@ def _get_page(self):
...
-            errors = GraphqlErrorSchema(many=True).load(
-                resp['error']['errors'], unknown=EXCLUDE)
+            errors = GraphqlErrorSchema(many=True).load(resp['error']['errors'], unknown=EXCLUDE)
             raise errors[0]
         if 'data' not in resp:
-            raise GraphqlParsingError(
-                'graphql data field was not returned but no error occurred')
+            raise GraphqlParsingError('graphql data field was not returned but no error occurred')
 
         connection_object = resp['data'][self._graph_object.object_name]
```
[the remark this answers is not in the document]
Comment (adamdelman-tenable) [reply tenable/ot/graphql/iterators.py]: fixed

File: tenable/ot/api.py [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+            :obj:`OTGraphIterator`:
+                An iterator object that will handle pagination of the data.
+
+        Example:
+            >>>     for event in tot.events.list(limit=500):
+                        print(event)
+        """
+        if not sort:
+            sort = [{"field": "eventId", "direction": "DescNullLast"}]
+            return super().list(EVENTS_QUERY, query_filter, search, sort, start_at, limit, **kwargs)
```
[the remark this answers is not in the document]
Comment (adamdelman-tenable) [reply tenable/ot/api.py]: fixed

File: tenable/ot/graphql/schema/assets.py [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+    '''
+    nodes = fields.List(fields.Nested(SegmentSchema))
+
+
+class RevisionSchema(Schema):
+    id = fields.UUID(required=True)
+    first_seen = fields.DateTime(required=True, data_key="firstSeen")
+    last_seen = fields.DateTime(required=True, data_key="lastSeen")
+    ordinal = fields.Integer(required=True)
+    is_base = fields.Boolean(required=True, data_key='isBase')
```
[the remark this answers is not in the document]
Comment (adamdelman-tenable) [reply tenable/ot/graphql/schema/assets.py]: fixed

File: tenable/ot/graphql/schema/events.py [hunk 9]
```diff
@@ -0,0 +1,208 @@
+from marshmallow import Schema, fields, post_load
+
+from tenable.ot.graphql.schema.assets import Asset
+from tenable.ot.schema.events import Event
+
+
```
[the remark this answers is not in the document]
Comment (adamdelman-tenable) [reply tenable/ot/graphql/schema/events.py]: fixed

File: tenable/ot/api.py (hunk 2 above)
Comment (adamdelman-tenable) [reply tenable/ot/api.py]: fixed

File: tenable/ot/api.py [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+
+        Returns:
+            :obj:`AnalysisResultsIterator`:
+                An iterator object handling data pagination.
+
+        Examples:
+            A quick example showing how to get the information for a specific
+            scan from SecurityCenter.  As the default is for the scan method to
+            return data from the vulndetails tool, we can handle this without
+            actually doing anything other than calling
```
[the remark this answers is not in the document]
Comment (adamdelman-tenable) [reply tenable/ot/api.py]: fixed

File: tenable/ot/graphql/definitions.py [hunk 11]
```diff
@@ -63,77 +65,105 @@ def __str__(self):
...
 
     @post_load
-    def to_object(self, data, **kwargs):
-        '''This method turns the schema into its corresponding object.'''
-        return Extensions(**data)
+    def to_object(self, data, *args, **kwargs):
+        """This method turns the schema into its corresponding object."""
+        if self.dataclass is None:
+            raise ValueError(f"dataclass must be set for {type(self)}")
+        normalized_data = {
```
Comment (AnavShapiraTenable) [inline tenable/ot/graphql/definitions.py:85]: not sure if .item() can failed, but maybe check that data is dict?

if type(data) is not dict:
      raise ValueError(f"data should be from type dict")

Comment (adamdelman-tenable) [reply tenable/ot/graphql/definitions.py:92]: done

File: tenable/ot/graphql/iterators.py [hunk 12]
```diff
@@ -1,51 +1,78 @@
...
+                "pageInfo"
+            ].get("endCursor", None)
+
+        return self.page
+
+    def _perform_request(
+        self,
+        graph_full_object: dict,
+    ) -> dict:
+        resp = self.api.graphql(**graph_full_object)
```
Comment (AnavShapiraTenable) [inline tenable/ot/graphql/iterators.py:61]: I think you should add timeout for request not responding (Ip not reachable and so on...)
Not sure if it should be configurable but maybe 10 sec? or maybe 20 sec?

File: tenable/ot/assets.py [hunk 13]
```diff
@@ -58,24 +59,10 @@ def list(self,
...
-        if filter:
-            query_variables['filter'] = filter
+            sort = [{"field": "id", "direction": "DescNullLast"}]
 
-        graph_obj = GraphObject(
-            object_name=ASSETS_QUERY_OBJECT_NAME,
-            query=ASSETS_QUERY,
-            resp_schema=AssetsSchema,
-            query_variables=query_variables
+        return super().list(
```
Comment (AnavShapiraTenable) [inline tenable/ot/assets.py:66]: Not sure if it is required, but maybe add support for assets.asset('asset_id') (very similay to  ot.plugins.plugin('Plugin_id') support) with built in query filter (I think it is very useful case...) WDYT?

File: tenable/ot/graphql/definitions.py [hunk 14]
```diff
@@ -63,77 +65,107 @@ def __str__(self):
...
+
+class GraphqlErrorSchema(Schema):
+    """
+    Schema for GraphQL error.
+    """
+
+    message = fields.Str(required=True)
+    locations = fields.List(fields.Nested(LocationSchema))
+    path = fields.List(fields.Str())
+    extensions = fields.Nested(ExtensionsSchema)
```
Comment (SteveMcGrath) [inline tenable/ot/graphql/definitions.py:166]: shouldn't locations and extensions be set to allow none as they are in an exception?

File: tenable/ot/graphql/iterators.py (hunk 12 above)
Comment (SteveMcGrath) [reply tenable/ot/graphql/iterators.py:61]: If a timeout is desired, then one should/count be set in the TenableOT class (which can be overloaded at instantiation) or can be passed to the .graphql method directly (using the timeout param)

https://restfly.readthedocs.io/en/latest/api/session.html#restfly.session.APISession._timeout

https://requests.readthedocs.io/en/latest/user/advanced/?highlight=timeout#timeouts

File: tenable/ot/graphql/query.py [hunk 15]
```diff
@@ -0,0 +1,423 @@
...
+          comment
+          owner
+          totalAffectedAssets
+        }
+      }
+    }
+  }
+}
+"""
+PLUGINS_QUERY = """
```
Comment (SteveMcGrath) [inline tenable/ot/graphql/query.py:131]: 2 blank lines.  comment detailing what this constant is

File: tenable/ot/graphql/query.py [hunk 16]
```diff
@@ -0,0 +1,423 @@
...
+        nodes {
+          id
+          name
+        }
+      }
+    }
+  }
+}
+"""
+EVENTS_QUERY = """
```
Comment (SteveMcGrath) [inline tenable/ot/graphql/query.py:235]: 2 blank lines.  comment detailing what this constant is

File: tenable/ot/graphql/schema/events.py [hunk 17]
```diff
@@ -0,0 +1,239 @@
...
+    archived = fields.Boolean(allow_none=True)
+    schema = fields.String(required=True)
+    continuous = fields.Boolean(required=True)
+    snapshot = fields.Boolean()
+    system = fields.Boolean(allow_none=True)
+    key = fields.String(required=True)
+    eventTypeDetails = fields.Nested(
+        EventTypeDetailsSchema, required=True, data_key="eventTypeDetails"
+    )
+    disableAfterHit = fields.Boolean(required=True, data_key="disableAfterHit")
```
Comment (SteveMcGrath) [inline tenable/ot/graphql/schema/events.py:138]: snake_case vs camelCase

File: tenable/ot/graphql/schema/plugins.py [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+    description = fields.String(required=True)
+    solution = fields.String(required=True)
+    see_also = fields.List(
+        fields.String(required=True), required=True, data_key="seeAlso"
+    )
+    plugin_type = fields.String(allow_none=True, data_key="pluginType")
+    plugin_pub_date = fields.String(allow_none=True, data_key="pluginPubDate")
+    plugin_mod_date = fields.String(allow_none=True, data_key="pluginModDate")
+    vuln_pub_date = fields.String(allow_none=True, data_key="vulnPubDate")
+    vuln_mod_date = fields.String(allow_none=True, data_key="vulnModDate")
```
Comment (SteveMcGrath) [inline tenable/ot/graphql/schema/plugins.py:31]: Shouldn't these be returned as datetime objects?

File: tenable/ot/plugins.py [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+
+        Example:
+            >>>     for plugin in tot.plugins.list(limit=500):
+                        print(plugin)
+        """
+        if not sort:
+            sort = [{"field": "id", "direction": "DescNullLast"}]
+
+        return super().list(
+            query, query_filter, search, sort, start_at, limit, **kwargs
```
Comment (SteveMcGrath) [inline tenable/ot/plugins.py:69]: Explicit keywords are preferred over implicit positional params

File: tenable/ot/schema/plugins.py [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+    source: str
+    family: str
+    description: str
+    solution: str
+    see_also: list[str]
+    plugin_type: str | None
+    plugin_pub_date: str | None
+    plugin_mod_date: str | None
+    vuln_pub_date: str | None
+    vuln_mod_date: str | None
```
Comment (SteveMcGrath) [inline tenable/ot/schema/plugins.py:26]: Shouldn't these be datetime?

File: tenable/ot/graphql/schema/events.py (hunk 17 above)
Comment (adamdelman-tenable) [reply tenable/ot/graphql/schema/events.py:138]: fixed

File: tenable/ot/graphql/schema/plugins.py (hunk 18 above)
Comment (adamdelman-tenable) [reply tenable/ot/graphql/schema/plugins.py:31]: fixed

File: tenable/ot/assets.py [hunk 21]
```diff
@@ -58,24 +59,10 @@ def list(self,
...
+            ASSETS_QUERY, query_filter, search, sort, start_at, limit, **kwargs
```
[the remark this answers is not in the document]
Comment (adamdelman-tenable) [reply tenable/ot/assets.py:67]: fixed

File: tenable/ot/graphql/definitions.py [hunk 22]
```diff
@@ -23,37 +25,37 @@ def __init__(self, object_name, query, resp_schema, query_variables):
 
     @property
     def object_name(self):
```
[the remark this answers is not in the document]
Comment (adamdelman-tenable) [reply tenable/ot/graphql/definitions.py:27]: fixed

File: tenable/ot/graphql/definitions.py [hunk 23]
```diff
@@ -23,37 +25,37 @@ def __init__(self, object_name, query, resp_schema, query_variables):
...
-        '''
+        """
         The name of the GraphQL object that appears in the query.
-        '''
+        """
         return self._object_name
 
     @property
     def query(self):
```
[the remark this answers is not in the document]
Comment (adamdelman-tenable) [reply tenable/ot/graphql/definitions.py:34]: fixed

File: tenable/ot/graphql/definitions.py [hunk 24]
```diff
@@ -23,37 +25,37 @@ def __init__(self, object_name, query, resp_schema, query_variables):
...
-        '''
+        """
         The GraphQL query.
-        '''
+        """
         return self._query
 
     @property
     def resp_schema(self):
```
[the remark this answers is not in the document]
Comment (adamdelman-tenable) [reply tenable/ot/graphql/definitions.py:41]: fixed

File: tenable/ot/graphql/definitions.py [hunk 25]
```diff
@@ -23,37 +25,37 @@ def __init__(self, object_name, query, resp_schema, query_variables):
...
-        '''
+        """
         The expected response schema from the query.
-        '''
+        """
         return self._resp_schema
 
     @property
     def query_variables(self):
```
[the remark this answers is not in the document]
Comment (adamdelman-tenable) [reply tenable/ot/graphql/definitions.py:48]: fixed

File: tenable/ot/events.py [hunk 26]
```diff
@@ -0,0 +1,67 @@
...
+
+        Example:
+            >>>     for event in tot.events.list(limit=500):
+                        print(event)
+        """
+        if not sort:
+            sort = [{"field": "eventId", "direction": "DescNullLast"}]
+
+        return super().list(
+            EVENTS_QUERY, query_filter, search, sort, start_at, limit, **kwargs
```
[the remark this answers is not in the document]
Comment (adamdelman-tenable) [reply tenable/ot/events.py:66]: fixed

File: tenable/ot/graphql/query.py [hunk 27]
```diff
@@ -0,0 +1,423 @@
+"""
+OT GraphQL API Queries.
+For each class, please refer to the Tenable.ot documentation website for a
+detailed explanation of the fields.
+"""
+
```
[the remark this answers is not in the document]
Comment (adamdelman-tenable) [reply tenable/ot/graphql/query.py:6]: fixed

File: tenable/ot/graphql/query.py (hunk 15 above)
Comment (adamdelman-tenable) [reply tenable/ot/graphql/query.py:131]: fixed

File: tenable/ot/graphql/query.py [hunk 28]
```diff
@@ -0,0 +1,423 @@
...
+        nodes {
+          id
+          name
+        }
+      }
+    }
+  }
+}
+"""
+PLUGINS_DETAILS_QUERY = """
```
[the remark this answers is not in the document]
Comment (adamdelman-tenable) [reply tenable/ot/graphql/query.py:169]: fixed

File: tenable/ot/graphql/query.py (hunk 16 above)
Comment (adamdelman-tenable) [reply tenable/ot/graphql/query.py:235]: fixed

File: tenable/ot/assets.py (hunk 13 above)
Comment (adamdelman-tenable) [reply tenable/ot/assets.py:66]: done

**Review by SteveMcGrath — CHANGES_REQUESTED** [review]

File: tenable/ot/graphql/definitions.py [hunk 29]
```diff
@@ -63,77 +37,102 @@ def __str__(self):
...
-    path = fields.List(fields.Str())
-    extensions = fields.Nested(ExtensionsSchema)
 
-    @post_load
-    def to_object(self, data, **kwargs):
-        '''This method turns the schema into its corresponding object.'''
-        return GraphqlError(**data)
+class NodesSchema(SchemaBase):
+    def __iter__(self):
+        yield from self.nodes
```
Comment (SteveMcGrath) [inline tenable/ot/graphql/definitions.py:68]: The iterable does not appear to have a test

**Comment by aseemsavio** [comment]
Hello @adamdelman-tenable, there seems to be a failing test.

```
FAILED tests/ot/test_events.py::test_list - AssertionError: assert Event(cate...
```

Thanks.

**Comment by aseemsavio** [comment]
@adamdelman-tenable there seems to be failing checks. Could you pl. merge master to your branch (and fix conflicts if any) and fix the failing checks?

Thanks.

**Comment by adamdelman-tenable** [comment]
Hi Aseem,

As far as I can see, there are no failing tests.
There seems to be a problem with the python 3.6 env setup in this job:
https://github.com/tenable/pyTenable/actions/runs/3306357872/jobs/5624719648

**Comment by aseemsavio** [comment]
@adamdelman-tenable Could you merge the latest changes from master into this branch? 

Thanks.

**Review by SteveMcGrath — APPROVED** [review]

**Pull request merged**